compiler/noirc_evaluator/src/ssa/opt/loop_invariant.rs

```diff
@@ -1521,4 +1868,264 @@ mod control_dependence {
+
+        let expected = "
+        brillig(inline) fn main f0 {
+          b0(v0: u32, v1: u32, v2: u32):
+            v4 = allocate -> &mut u32
+            store v0 at v4
+            v6 = eq v1, u32 4
+            jmp b1(u32 0)
+          b1(v3: u32):
+            v9 = lt v3, u32 5
+            jmpif v9 then: b2, else: b3
+          b2():
+            jmpif u1 1 then: b4, else: b5
+          b3():
+            v19 = load v4 -> u32
+            v20 = lt v1, v19
+            constrain v20 == u1 1
+            return
+          b4():
+            v11 = load v4 -> u32
+            v13 = add v11, u32 1
+            store v13 at v4
+            jmp b5()
+          b5():
+            v15 = lt u32 2, v3
+            v16 = mul v6, v15
+            jmpif v16 then: b3, else: b6
+          b6():
+            v17 = lt v3, u32 4
+            constrain v17 == u1 1
+            v18 = unchecked_add v3, u32 1
+            jmp b1(v18)
+        }
+        ";
+
+        assert_normalized_ssa_equals(ssa, expected);
+    }
+
+    #[test]
+    fn simplify_comparison() {
+        // This tests shows that the comparison v12 = lt v3, u32 8 is simplified because v3 is bounded by 5
+        let src = "
+        brillig(inline) fn main f0 {
+          b0(v0: u32, v1: u32, v2: u32):
+            v4 = allocate -> &mut u32
+            store v0 at v4
+            jmp b1(u32 0)
+          b1(v3: u32):
+            v7 = lt v3, u32 5
+            jmpif v7 then: b2, else: b3
+          b2():
+            v12 = lt v3, u32 8
+            jmpif v12 then: b4, else: b5
+          b3():
+            v8 = load v4 -> u32
+            v9 = lt v1, v8
+            constrain v9 == u1 1
+            return
+          b4():
+            v13 = load v4 -> u32
+            v15 = add v13, u32 1
+            store v15 at v4
+            jmp b5()
+          b5():
+            v16 = unchecked_add v3, u32 1
+            jmp b1(v16)
+        }
+        ";
+
+        let ssa = Ssa::from_str(src).unwrap();
+
+        let ssa = ssa.loop_invariant_code_motion();
+
+        let expected = "
+        brillig(inline) fn main f0 {
+          b0(v0: u32, v1: u32, v2: u32):
+            v4 = allocate -> &mut u32
+            store v0 at v4
+            jmp b1(u32 0)
+          b1(v3: u32):
+            v7 = lt v3, u32 5
+            jmpif v7 then: b2, else: b3
+          b2():
+            jmpif u1 1 then: b4, else: b5
+          b3():
+            v8 = load v4 -> u32
+            v9 = lt v1, v8
+            constrain v9 == u1 1
+            return
+          b4():
+            v11 = load v4 -> u32
+            v13 = add v11, u32 1
+            store v13 at v4
+            jmp b5()
+          b5():
+            v14 = unchecked_add v3, u32 1
+            jmp b1(v14)
+        }
+        ";
+
+        assert_normalized_ssa_equals(ssa, expected);
+    }
+
+    #[test]
+    fn simplify_not_equal_constrain() {
+        // This tests shows that the not equal constraint on v3 is simplified due to the loop bounds
```

Comment:
I would prefer we say "not equal instruction block" or something instead, as this SSA is relevant for Brillig as well where there are not constraints and `constrain` is simply an instruction in our SSA. I noticed this in a couple other comments in the tests, I think "instruction" over "constraint" is more accurate.

Reply:
I changed it but kept the other one. I agree this is relevant to brillig but the test is only for ssa, as well as the loop invariant pass. The constrain is the reason why we are able to do the optimisation tested here and in do_not_simplify_constraint() test.

Reply:
Oh I just meant in the comment not necessarily the test name. Reading the tests against let's just go back to what you had with `simplfiy_not_equal_constraint`. Apologies for the confusion.